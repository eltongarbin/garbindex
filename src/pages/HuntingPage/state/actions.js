import { createAction } from 'redux-actions';

import {
  createType,
  createAsyncTypes,
  createAsyncActions
} from 'utils/reduxHelpers';

export const types = {
  CLEAN_RESULT: createType('pokedex', 'CLEAN_RESULT'),
  SEARCH_POKEMON: createAsyncTypes('pokedex', 'SEARCH_POKEMON')
};

export default {
  cleanSearchResult: createAction(types.CLEAN_RESULT),
  searchForPokemon: createAsyncActions(types.SEARCH_POKEMON)
};
