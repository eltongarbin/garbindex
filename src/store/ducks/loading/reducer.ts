import { LoadingState } from './types';
import { Action } from 'typesafe-actions';

const initialState: LoadingState = {};

export default (state = initialState, { type }: Action) => {
  const matches = /(.*)_(REQUEST|SUCCESS|FAILURE)/.exec(type);

  if (!matches) return state;

  const [, requestName, requestState] = matches;
  return {
    ...state,
    [requestName]: requestState === 'REQUEST'
  };
};
