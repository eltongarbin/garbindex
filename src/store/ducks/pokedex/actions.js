import { createAction } from 'redux-actions';
import { createType } from 'utils/reduxHelpers';

export const types = {
  RELEASE_BYID: createType('pokedex', 'RELEASE_BYID'),
  CATCH_BYID: createType('pokedex', 'CATCH_BYID'),
  CHANGE_IMAGE: createType('pokedex', 'CHANGE_IMAGE')
};

export default {
  releasePokemon: createAction(types.RELEASE_BYID),
  catchPokemon: createAction(types.CATCH_BYID),
  changePokemonImage: createAction(types.CHANGE_IMAGE)
};
