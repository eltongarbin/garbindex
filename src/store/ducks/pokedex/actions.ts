import { createAction } from 'typesafe-actions';
import { ActionTypes } from './types';

export const releasePokemon = createAction(ActionTypes.RELEASE_BYID)<number>();
export const catchPokemon = createAction(ActionTypes.CATCH_BYID)<number>();
export const changePokemonImage = createAction(ActionTypes.CHANGE_IMAGE)<{
  id: number;
  image: string;
}>();
