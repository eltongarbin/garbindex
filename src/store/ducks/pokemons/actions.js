import { createAsyncTypes, createAsyncActions } from 'utils/reduxHelpers';

export const types = {
  FETCH_POKEMON: createAsyncTypes('pokemons', 'FETCH_POKEMON'),
  FETCH_EVOLUTION_CHAIN: createAsyncTypes('pokemons', 'FETCH_EVOLUTION_CHAIN')
};

export default {
  fetchPokemon: createAsyncActions(types.FETCH_POKEMON),
  fetchEvolutionChain: createAsyncActions(types.FETCH_EVOLUTION_CHAIN)
};
