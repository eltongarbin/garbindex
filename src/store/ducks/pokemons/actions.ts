import { createAsyncAction } from 'typesafe-actions';

import { ActionTypes } from './types';

export const fetchPokemon = createAsyncAction(
  ActionTypes.FETCH_POKEMON_REQUEST,
  ActionTypes.FETCH_POKEMON_SUCCESS, // TODO: Improve payload type
  ActionTypes.FETCH_POKEMON_FAILURE
)<string | number, any, Error>();

export const fetchEvolvedFrom = createAsyncAction(
  ActionTypes.FETCH_EVOLUTION_CHAIN_REQUEST,
  ActionTypes.FETCH_EVOLUTION_CHAIN_SUCCESS, // TODO: Improve payload type
  ActionTypes.FETCH_EVOLUTION_CHAIN_FAILURE
)<number, any, Error>();

export const fetchPokemonsByTypeId = createAsyncAction(
  ActionTypes.FETCH_POKEMONS_BYTYPE_REQUEST, // TODO: Improve payload type
  ActionTypes.FETCH_POKEMONS_BYTYPE_SUCCESS, // TODO: Improve payload type
  ActionTypes.FETCH_POKEMONS_BYTYPE_FAILURE
)<{ pokemonId: number; typeId: number }, any, Error>();

export const fetchShortEffectByAbilityId = createAsyncAction(
  ActionTypes.FETCH_SHORT_EFFECT_BYABILITY_REQUEST, // TODO: Improve payload type
  ActionTypes.FETCH_SHORT_EFFECT_BYABILITY_SUCCESS, // TODO: Improve payload type
  ActionTypes.FETCH_SHORT_EFFECT_BYABILITY_FAILURE
)<{ pokemonId: number; abilityId: number }, any, Error>();
