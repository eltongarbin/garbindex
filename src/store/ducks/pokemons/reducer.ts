import { createReducer } from 'typesafe-actions';
import { keyBy } from 'lodash-es';
import { extractParamId } from 'utils';
import { DeepReadonly } from 'utility-types';

import {
  fetchPokemon,
  fetchEvolvedFrom,
  fetchPokemonsByTypeId,
  fetchShortEffectByAbilityId
} from './actions';

export type PokemonType = {
  id: number;
  name: string;
  pokemons?: Array<{
    id: number;
    name: string;
  }>;
};

export type PokemonAbility = {
  id: number;
  name: string;
  short_effect?: string;
};

type PokemonState = DeepReadonly<{
  byId: IDictionary<{
    id: number;
    name: string;
    image: string;
    evolvedFrom?: string;
    height: number;
    weight: number;
    stats: Array<{
      id: number;
      name: string;
      base_stat: number;
    }>;
    abilitiesById: IDictionary<PokemonAbility>;
    typesById: IDictionary<PokemonType>;
  }>;
}>;

const initialState: PokemonState = {
  byId: {}
};

export default createReducer(initialState)
  .handleAction(fetchPokemon.success, (state, { payload }) => ({
    byId: {
      ...state.byId,
      [payload.id]: {
        ...payload,
        image: payload.sprites.front_default,
        abilitiesById: keyBy(
          payload.abilities.map(({ ability }) => ({
            id: parseInt(extractParamId(ability.url)),
            name: ability.name
          })),
          'id'
        ),
        stats: payload.stats.map(({ base_stat, stat }) => ({
          base_stat: base_stat,
          name: stat.name,
          id: parseInt(extractParamId(stat.url))
        })),
        typesById: keyBy(
          payload.types.map(({ type }) => ({
            id: parseInt(extractParamId(type.url)),
            name: type.name
          })),
          'id'
        )
      }
    }
  }))
  .handleAction(fetchEvolvedFrom.success, (state, { payload }) => ({
    byId: {
      ...state.byId,
      [payload.id]: {
        ...state.byId[payload.id],
        ...payload,
        evolvedFrom: payload.evolves_from_species
          ? payload.evolves_from_species.name
          : null
      }
    }
  }))
  .handleAction(fetchPokemonsByTypeId.success, (state, { payload }) => {
    const currentPokemon = state.byId[payload.pokemonId];
    const currentType = currentPokemon.typesById[payload.typeId];

    return {
      ...state,
      byId: {
        ...state.byId,
        [currentPokemon.id]: {
          ...currentPokemon,
          typesById: {
            ...currentPokemon.typesById,
            [currentType.id]: { ...currentType, pokemons: payload.pokemons }
          }
        }
      }
    };
  })
  .handleAction(fetchShortEffectByAbilityId.success, (state, { payload }) => {
    const currentPokemon = state.byId[payload.pokemonId];
    const currentAbility = currentPokemon.abilitiesById[payload.abilityId];

    return {
      ...state,
      byId: {
        ...state.byId,
        [currentPokemon.id]: {
          ...currentPokemon,
          abilitiesById: {
            ...currentPokemon.abilitiesById,
            [currentAbility.id]: {
              ...currentAbility,
              short_effect: payload.short_effect
            }
          }
        }
      }
    };
  });
