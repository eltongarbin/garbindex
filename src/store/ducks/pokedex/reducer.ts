import { handleActions } from 'redux-actions';
import { types } from './actions';

const initialState = {
  pokemonsId: [],
  pokemonsCustomizedById: {}
};

export default handleActions(
  {
    [types.RELEASE_BYID]: (state, { payload }) => {
      const { [payload]: value, ...rest } = state.pokemonsCustomizedById;

      return {
        pokemonsId: state.pokemonsId.filter((id) => id !== payload),
        pokemonsCustomizedById: rest
      };
    },
    [types.CATCH_BYID]: (state, { payload }) => ({
      ...state,
      pokemonsId: [...state.pokemonsId, payload]
    }),
    [types.CHANGE_IMAGE]: (state, { payload }) => ({
      ...state,
      pokemonsCustomizedById: {
        ...state.pokemonsCustomizedById,
        [payload.id]: {
          ...state.pokemonsCustomizedById[payload.id],
          ...payload
        }
      }
    })
  },
  initialState
);
