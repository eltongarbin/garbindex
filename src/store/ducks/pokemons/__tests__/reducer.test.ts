import * as actions from '../actions';
import reducer, { PokemonState } from '../reducer';

const getInitialState = (initial?: Partial<PokemonState>) =>
  reducer(initial as PokemonState, {} as any);

it('should return the initial state by default', () => {
  const initialState = getInitialState();
  expect(initialState).toMatchInlineSnapshot(`
    Object {
      "byId": Object {},
    }
  `);
});

it('should add pokemon searched', () => {
  const initialState = getInitialState();
  const state = reducer(
    initialState,
    actions.fetchPokemon.success({
      abilities: [
        {
          ability: {
            name: 'water-veil',
            url: 'https://pokeapi.co/api/v2/ability/41/'
          }
        },
        {
          ability: {
            name: 'storm-drain',
            url: 'https://pokeapi.co/api/v2/ability/114/'
          }
        },
        {
          ability: {
            name: 'swift-swim',
            url: 'https://pokeapi.co/api/v2/ability/33/'
          }
        }
      ],
      height: 4,
      id: 456,
      name: 'finneon',
      sprites: {
        front_default:
          'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/456.png'
      },
      stats: [
        {
          base_stat: 66,
          stat: {
            name: 'speed',
            url: 'https://pokeapi.co/api/v2/stat/6/'
          }
        },
        {
          base_stat: 61,
          stat: {
            name: 'special-defense',
            url: 'https://pokeapi.co/api/v2/stat/5/'
          }
        },
        {
          base_stat: 49,
          stat: {
            name: 'special-attack',
            url: 'https://pokeapi.co/api/v2/stat/4/'
          }
        },
        {
          base_stat: 56,
          stat: {
            name: 'defense',
            url: 'https://pokeapi.co/api/v2/stat/3/'
          }
        },
        {
          base_stat: 49,
          stat: {
            name: 'attack',
            url: 'https://pokeapi.co/api/v2/stat/2/'
          }
        },
        {
          base_stat: 49,
          stat: {
            name: 'hp',
            url: 'https://pokeapi.co/api/v2/stat/1/'
          }
        }
      ],
      types: [
        {
          type: {
            name: 'water',
            url: 'https://pokeapi.co/api/v2/type/11/'
          }
        }
      ],
      weight: 70
    })
  );
  expect(state).toMatchSnapshot();
});
