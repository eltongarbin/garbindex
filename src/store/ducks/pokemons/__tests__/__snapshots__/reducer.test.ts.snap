// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should add pokemon searched 1`] = `
Object {
  "byId": Object {
    "456": Object {
      "abilities": Array [
        Object {
          "ability": Object {
            "name": "water-veil",
            "url": "https://pokeapi.co/api/v2/ability/41/",
          },
        },
        Object {
          "ability": Object {
            "name": "storm-drain",
            "url": "https://pokeapi.co/api/v2/ability/114/",
          },
        },
        Object {
          "ability": Object {
            "name": "swift-swim",
            "url": "https://pokeapi.co/api/v2/ability/33/",
          },
        },
      ],
      "abilitiesById": Object {
        "114": Object {
          "id": 114,
          "name": "storm-drain",
        },
        "33": Object {
          "id": 33,
          "name": "swift-swim",
        },
        "41": Object {
          "id": 41,
          "name": "water-veil",
        },
      },
      "height": 4,
      "id": 456,
      "image": "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/456.png",
      "name": "finneon",
      "sprites": Object {
        "front_default": "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/456.png",
      },
      "stats": Array [
        Object {
          "base_stat": 66,
          "id": 6,
          "name": "speed",
        },
        Object {
          "base_stat": 61,
          "id": 5,
          "name": "special-defense",
        },
        Object {
          "base_stat": 49,
          "id": 4,
          "name": "special-attack",
        },
        Object {
          "base_stat": 56,
          "id": 3,
          "name": "defense",
        },
        Object {
          "base_stat": 49,
          "id": 2,
          "name": "attack",
        },
        Object {
          "base_stat": 49,
          "id": 1,
          "name": "hp",
        },
      ],
      "types": Array [
        Object {
          "type": Object {
            "name": "water",
            "url": "https://pokeapi.co/api/v2/type/11/",
          },
        },
      ],
      "typesById": Object {
        "11": Object {
          "id": 11,
          "name": "water",
        },
      },
      "weight": 70,
    },
  },
}
`;
